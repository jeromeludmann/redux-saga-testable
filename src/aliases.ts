import {
  take,
  takeMaybe,
  takeEvery,
  takeLatest,
  takeLeading,
  put,
  putResolve,
  call,
  apply,
  cps,
  fork,
  spawn,
  join,
  cancel,
  select,
  actionChannel,
  flush,
  cancelled,
  setContext,
  getContext,
  delay,
  throttle,
  debounce,
  retry,
  all,
  race,
} from 'redux-saga/effects'
import { Effect } from '@redux-saga/types'
import { SagaRunner } from './types/runner'

export const getExtendedSagaAssertions = (
  runner: SagaRunner,
  isNegated: () => boolean,
  _yield: (
    runner: SagaRunner,
    isNegated: () => boolean,
  ) => (...args: any[]) => SagaRunner,
) => {
  const createAlias = (effectCreator: (...args: any[]) => Effect) => (
    ...args: any[]
  ) => _yield(runner, isNegated)(effectCreator(...args))

  return {
    take: createAlias(take),
    takeMaybe: createAlias(takeMaybe),
    takeEvery: createAlias(takeEvery),
    takeLatest: createAlias(takeLatest),
    takeLeading: createAlias(takeLeading),
    put: createAlias(put),
    putResolve: createAlias(putResolve),
    call: createAlias(call),
    apply: createAlias(apply),
    cps: createAlias(cps),
    fork: createAlias(fork),
    spawn: createAlias(spawn),
    join: createAlias(join),
    cancel: createAlias(cancel),
    select: createAlias(select),
    actionChannel: createAlias(actionChannel),
    flush: createAlias(flush),
    cancelled: createAlias(cancelled),
    setContext: createAlias(setContext),
    getContext: createAlias(getContext),
    delay: createAlias(delay),
    throttle: createAlias(throttle),
    debounce: createAlias(debounce),
    retry: createAlias(retry),
    all: createAlias(all),
    race: createAlias(race),
  }
}
