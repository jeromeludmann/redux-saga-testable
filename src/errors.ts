import { stringify } from './strings';

export class RunnerError extends Error {
  constructor(message: string, params: any[] = [], fn?: Function) {
    super(
      params.reduce((message, param) => {
        if (typeof param !== 'string') {
          param = stringify(param);
        }
        message += `\n\n${param}`;
        return message;
      }, message),
    );

    this.name = RunnerError.name;
    Error.captureStackTrace(this, fn ?? RunnerError);
  }
}

export function setCallSite(error: Error, fn: Function) {
  if (error.name === RunnerError.name) {
    Error.captureStackTrace(error, fn);
  }
}
