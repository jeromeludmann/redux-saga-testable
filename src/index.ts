import { _createRunner } from './runner'
import { createError } from './utils'
import {
  SagaRunner,
  ThrowError,
  THROW_ERROR,
  Finalize,
  FINALIZE,
} from './types/runner'

/**
 * Creates a saga runner.
 */
export function createRunner<Saga extends (...args: any[]) => any>(
  saga: Saga,
  ...args: Parameters<Saga>
): SagaRunner {
  if (!saga) {
    throw createError('Missing saga argument', createRunner)
  }

  return _createRunner(
    {
      injections: [],
      assertions: [],
    },
    saga,
    args,
  )
}

/**
 * Creates a saga runner.
 * @deprecated Use `createRunner()` instead.
 */
export const use = createRunner

/**
 * Throws an error from the saga when injected as a value.
 */
export function throwError(error: Error): ThrowError {
  return {
    [THROW_ERROR]: true,
    error,
  }
}

/**
 * Finalizes the saga when injected as a value.
 */
export function finalize(): Finalize {
  return {
    [FINALIZE]: true,
  }
}
