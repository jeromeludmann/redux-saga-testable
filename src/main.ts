export { createRunner, Runner } from './runner';

export {
  throwError,
  finalize,
  RunnerOutput,
  ThrowError,
  Finalize,
} from './engine';
