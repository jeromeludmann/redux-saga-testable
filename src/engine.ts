import { isDeepStrictEqual } from 'util';
import { Effect, Saga } from '@redux-saga/types';

import { RunnerError } from './errors';
import { stringify } from './strings';
import { ErrorPattern, isEffect, matchError } from './utils';

/**
 * The runner output returned by `runner.run()`.
 */
export interface RunnerOutput {
  effects: Effect[];
  return?: unknown;
  error?: Error;
}

interface EffectMapping {
  effect: Effect;
  values: unknown[];
}

interface InitialState {
  mappings: EffectMapping[];
  errorToCatch?: ErrorPattern;
}

const ENGINE = '@@redux-saga-testable/engine';

const THROW_ERROR = 'THROW_ERROR';

export interface ThrowError {
  [ENGINE]: typeof THROW_ERROR;
  error: Error;
}

const FINALIZE = 'FINALIZE';

export interface Finalize<T> {
  [ENGINE]: typeof FINALIZE;
  value?: T;
}

const MAX_YIELDED_EFFECTS = 100;

export class Engine {
  protected readonly saga: Saga;
  protected readonly args: Parameters<Saga>;
  protected mappings: EffectMapping[];
  protected errorToCatch?: ErrorPattern;
  private cachedOutput: RunnerOutput | null;

  constructor(saga: Saga, args: unknown[], initialState?: InitialState) {
    this.saga = saga;
    this.args = args;
    this.mappings = initialState?.mappings ?? [];
    this.errorToCatch = initialState?.errorToCatch;
    this.cachedOutput = null;
  }

  /**
   * Maps an effect to a value.
   */
  map(effect: Effect, value: unknown, ...nextValues: unknown[]): this {
    if (arguments.length < 1) {
      throw new RunnerError('Missing effect argument', this.map);
    }

    if (arguments.length < 2) {
      throw new RunnerError(
        ['The value to map is missing', 'Given effect:', stringify(effect)],
        this.map,
      );
    }

    const existingMapping = this.mappings.find(mapping =>
      isDeepStrictEqual(mapping.effect, effect),
    );

    if (existingMapping) {
      throw new RunnerError(
        [
          'Mapped values already provided for this effect',
          'Given effect:',
          stringify(effect),
          'Existing mapped values:',
          stringify(existingMapping.values),
        ],
        this.map,
      );
    }

    this.mappings.push({
      effect,
      values: [value, ...nextValues],
    });

    this.cachedOutput = null;

    return this;
  }

  /**
   * Catches silently an error thrown by the saga.
   */
  catch(error: ErrorPattern): this {
    if (arguments.length < 1) {
      throw new RunnerError('Missing error pattern argument', this.catch);
    }

    if (this.errorToCatch) {
      throw new RunnerError(
        [
          'Error pattern already provided',
          'Given error pattern:',
          stringify(this.errorToCatch),
        ],
        this.catch,
      );
    }

    this.errorToCatch = error;
    this.cachedOutput = null;

    return this;
  }

  /**
   * Runs the saga.
   */
  run(): RunnerOutput {
    // If an output already exists for the current engine state,
    // does not rerun the saga and returns the existing cached output.
    if (this.cachedOutput !== null) {
      return this.cachedOutput;
    }

    this.cachedOutput = { effects: [] };

    const iterator = this.saga(...this.args);
    let step: IteratorResult<any>;
    let nextValue = undefined;

    // Allows the runner to consume the mapping values without mutating them.
    const mappings: EffectMapping[] = this.mappings.map(mapping => ({
      ...mapping,
      values: Array.from(mapping.values),
    }));

    // Runs the saga, breaks when is done or when an error is thrown.
    for (;;) {
      try {
        step = this.next(iterator, nextValue);
      } catch (sagaError) {
        this.cachedOutput.error = sagaError;
        break;
      }

      if (step.done) {
        this.cachedOutput.return = step.value;
        break;
      }

      if (!isEffect(step.value)) {
        nextValue = step.value;
        continue;
      }

      if (this.cachedOutput.effects.length > MAX_YIELDED_EFFECTS) {
        throw new RunnerError('Too many yielded effects', this.run);
      }

      this.cachedOutput.effects.push(step.value);

      // Gets the value mapped with the current effect
      // and uses it as the next value to be passed to the generator.
      nextValue = mappings
        .find(mapping => isDeepStrictEqual(mapping.effect, step.value))
        ?.values.shift();
    }

    const unusedMapping = mappings.find(mapping => mapping.values.length > 0);

    if (unusedMapping) {
      throw new RunnerError(
        [
          'Unused mapped values',
          'Given effect:',
          stringify(unusedMapping.effect),
          'Unused mapped values:',
          stringify(unusedMapping.values),
        ],
        this.run,
      );
    }

    // Rethrows the error thrown by the saga itself.
    if (this.cachedOutput.error) {
      if (
        !this.errorToCatch ||
        !matchError(this.cachedOutput.error, this.errorToCatch)
      ) {
        throw this.cachedOutput.error;
      }
    }

    // Checks for useless `runner.catch()`.
    else if (this.errorToCatch) {
      throw new RunnerError(
        [
          'No error thrown by the saga',
          'Given error pattern:',
          stringify(this.errorToCatch),
        ],
        this.run,
      );
    }

    return this.cachedOutput;
  }

  private next<T>(iterator: Iterator<T>, value: any): IteratorResult<any> {
    if (value !== null && typeof value === 'object' && ENGINE in value) {
      switch (value[ENGINE]) {
        case THROW_ERROR:
          return iterator.throw!(value.error);
        case FINALIZE:
          if (value.value !== undefined) {
            iterator.next(value.value);
          }
          return iterator.return!();
      }
    }

    return iterator.next(value);
  }
}

/**
 * Throws an error from the saga when mapped as a value.
 */
export function throwError(error: Error): ThrowError {
  if (arguments.length < 1) {
    throw new RunnerError('Missing error argument', throwError);
  }

  return {
    [ENGINE]: THROW_ERROR,
    error,
  };
}

/**
 * Finalizes the saga when mapped as a value.
 */
export function finalize<T>(value?: T): Finalize<T> {
  return {
    [ENGINE]: FINALIZE,
    value,
  };
}
