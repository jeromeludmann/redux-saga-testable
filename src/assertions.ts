import { isDeepStrictEqual } from 'util';
import { Action } from 'redux';
import { TakeableChannel, PuttableChannel, FlushableChannel } from 'redux-saga';
import { Effect } from 'redux-saga/effects';
import * as effects from 'redux-saga/effects';
import * as types from '@redux-saga/types';

import { Engine } from './engine';
import { RunnerError, setCallSite } from './errors';
import { ErrorPattern, matchError } from './utils';

export class Assertions<R extends Engine> {
  private readonly runner: R;
  private negated: boolean;

  constructor({ runner }: { runner: R }) {
    this.runner = runner;
    this.negated = false;

    this.take = this.createAlias(effects.take);
    this.takeMaybe = this.createAlias(effects.takeMaybe);
    this.takeEvery = this.createAlias(effects.takeEvery);
    this.takeLatest = this.createAlias(effects.takeLatest);
    this.takeLeading = this.createAlias(effects.takeLeading);
    this.put = this.createAlias(effects.put);
    this.putResolve = this.createAlias(effects.putResolve);
    this.call = this.createAlias(effects.call);
    this.apply = this.createAlias(effects.apply);
    this.cps = this.createAlias(effects.cps);
    this.fork = this.createAlias(effects.fork);
    this.spawn = this.createAlias(effects.spawn);
    this.join = this.createAlias(effects.join);
    this.cancel = this.createAlias(effects.cancel);
    this.select = this.createAlias(effects.select);
    this.actionChannel = this.createAlias(effects.actionChannel);
    this.flush = this.createAlias(effects.flush);
    this.cancelled = this.createAlias(effects.cancelled);
    this.setContext = this.createAlias(effects.setContext);
    this.getContext = this.createAlias(effects.getContext);
    this.delay = this.createAlias(effects.delay);
    this.throttle = this.createAlias(effects.throttle);
    this.debounce = this.createAlias(effects.debounce);
    this.retry = this.createAlias(effects.retry);
    this.all = this.createAlias(effects.all);
    this.race = this.createAlias(effects.race);
  }

  /**
   * Asserts that the saga yields an effect.
   */
  yield(effect: Effect) {
    try {
      if (!effect) {
        throw new RunnerError('Missing effect argument');
      }

      const output = this.runner.run();

      if (
        !this.assert(output.effects.some(e => isDeepStrictEqual(e, effect)))
      ) {
        throw new RunnerError('Assertion failure', [
          `${this.expected} effect:`,
          effect,
          'Received effects:',
          output.effects,
        ]);
      }

      return this.runner;
    } catch (error) {
      setCallSite(error, this.yield);
      throw error;
    } finally {
      this.negated = false;
    }
  }

  /**
   * Asserts that the saga returns a value.
   */
  return<T>(value: T) {
    try {
      if (arguments.length < 1) {
        throw new RunnerError('Missing value argument');
      }

      const output = this.runner.run();

      if (!this.assert(isDeepStrictEqual(output.return, value))) {
        throw new RunnerError('Assertion failure', [
          `${this.expected} return value:`,
          value,
          'Received return value:',
          output.return,
        ]);
      }

      return this.runner;
    } catch (error) {
      setCallSite(error, this.return);
      throw error;
    } finally {
      this.negated = false;
    }
  }

  /**
   * Asserts that the saga throws an error.
   */
  throw(error: ErrorPattern) {
    try {
      if (!error) {
        throw new RunnerError('Missing error pattern argument');
      }

      const output = this.runner.run();

      if (!this.assert(!!output.error && matchError(output.error, error))) {
        throw new RunnerError('Assertion failure', [
          `${this.expected} error pattern:`,
          error,
          'Received thrown error:',
          output.error,
        ]);
      }

      return this.runner;
    } catch (error) {
      setCallSite(error, this.throw);
      throw error;
    } finally {
      this.negated = false;
    }
  }

  /**
   * Negates the next assertion.
   */
  get not(): Omit<Assertions<R>, 'not'> {
    this.negated = true;
    return this;
  }

  private assert(assertion: boolean) {
    return this.negated ? !assertion : assertion;
  }

  private get expected() {
    return this.negated ? 'Not expected' : 'Expected';
  }

  private createAlias(effectCreator: (...args: any[]) => Effect) {
    const alias = (...effectArgs: unknown[]) => {
      try {
        return this.yield(effectCreator(...effectArgs));
      } catch (error) {
        setCallSite(error, alias);
        throw error;
      }
    };

    return alias;
  }

  // Extended assertions.
  // Based on the redux-saga type definitions.
  // See https://github.com/jeromeludmann/redux-saga-testable/issues/31

  /**
   * Alias for `runner.should.yield(take(...))`
   */
  take: {
    (pattern?: types.ActionPattern): R;

    <A extends Action>(pattern?: types.ActionPattern<A>): R;

    <T>(channel: TakeableChannel<T>, multicastPattern?: types.Pattern<T>): R;
  };

  /**
   * Alias for `runner.should.yield(takeMaybe(...))`
   */
  takeMaybe: {
    (pattern?: types.ActionPattern): R;

    <A extends Action>(pattern?: types.ActionPattern<A>): R;

    <T>(channel: TakeableChannel<T>, multicastPattern?: types.Pattern<T>): R;
  };

  /**
   * Alias for `runner.should.yield(takeEvery(...))`
   */
  takeEvery: {
    <P extends types.ActionPattern>(
      pattern: P,
      worker: (action: types.ActionMatchingPattern<P>) => any,
    ): R;

    <P extends types.ActionPattern, Fn extends (...args: any[]) => any>(
      pattern: P,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<
        types.ActionMatchingPattern<P>,
        Fn
      >
    ): R;

    <A extends Action>(
      pattern: types.ActionPattern<A>,
      worker: (action: A) => any,
    ): R;

    <A extends Action, Fn extends (...args: any[]) => any>(
      pattern: types.ActionPattern<A>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<A, Fn>
    ): R;

    <T>(channel: TakeableChannel<T>, worker: (item: T) => any): R;

    <T, Fn extends (...args: any[]) => any>(
      channel: TakeableChannel<T>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<T, Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(takeLatest(...))`
   */
  takeLatest: {
    <P extends types.ActionPattern>(
      pattern: P,
      worker: (action: types.ActionMatchingPattern<P>) => any,
    ): R;

    <P extends types.ActionPattern, Fn extends (...args: any[]) => any>(
      pattern: P,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<
        types.ActionMatchingPattern<P>,
        Fn
      >
    ): R;

    <A extends Action>(
      pattern: types.ActionPattern<A>,
      worker: (action: A) => any,
    ): R;

    <A extends Action, Fn extends (...args: any[]) => any>(
      pattern: types.ActionPattern<A>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<A, Fn>
    ): R;

    <T>(channel: TakeableChannel<T>, worker: (item: T) => any): R;

    <T, Fn extends (...args: any[]) => any>(
      channel: TakeableChannel<T>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<T, Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(takeLeading(...))`
   */
  takeLeading: {
    <P extends types.ActionPattern>(
      pattern: P,
      worker: (action: types.ActionMatchingPattern<P>) => any,
    ): R;

    <P extends types.ActionPattern, Fn extends (...args: any[]) => any>(
      pattern: P,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<
        types.ActionMatchingPattern<P>,
        Fn
      >
    ): R;

    <A extends Action>(
      pattern: types.ActionPattern<A>,
      worker: (action: A) => any,
    ): R;

    <A extends Action, Fn extends (...args: any[]) => any>(
      pattern: types.ActionPattern<A>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<A, Fn>
    ): R;

    <T>(channel: TakeableChannel<T>, worker: (item: T) => any): R;

    <T, Fn extends (...args: any[]) => any>(
      channel: TakeableChannel<T>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<T, Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(put(...))`
   */
  put: {
    <A extends Action>(action: A): R;

    <T>(channel: PuttableChannel<T>, action: T | types.END): R;
  };

  /**
   * Alias for `runner.should.yield(putResolve(...))`
   */
  putResolve: <A extends Action>(action: A) => R;

  /**
   * Alias for `runner.should.yield(call(...))`
   */
  call: {
    <Fn extends (...args: any[]) => any>(fn: Fn, ...args: Parameters<Fn>): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctxAndFnName: [Ctx, Name],
      ...args: Parameters<Ctx[Name]>
    ): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctxAndFnName: { context: Ctx; fn: Name },
      ...args: Parameters<Ctx[Name]>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctxAndFn: [Ctx, Fn],
      ...args: Parameters<Fn>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctxAndFn: { context: Ctx; fn: Fn },
      ...args: Parameters<Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(apply(...))`
   */
  apply: {
    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctx: Ctx,
      fnName: Name,
      args: Parameters<Ctx[Name]>,
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctx: Ctx,
      fn: Fn,
      args: Parameters<Fn>,
    ): R;
  };

  /**
   * Alias for `runner.should.yield(cps(...))`
   */
  cps: {
    <Fn extends (cb: effects.CpsCallback<any>) => any>(fn: Fn): R;

    <Fn extends (...args: any[]) => any>(
      fn: Fn,
      ...args: effects.CpsFunctionParameters<Fn>
    ): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => void },
      Name extends string
    >(
      ctxAndFnName: [Ctx, Name],
      ...args: effects.CpsFunctionParameters<Ctx[Name]>
    ): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => void },
      Name extends string
    >(
      ctxAndFnName: { context: Ctx; fn: Name },
      ...args: effects.CpsFunctionParameters<Ctx[Name]>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => void>(
      ctxAndFn: [Ctx, Fn],
      ...args: effects.CpsFunctionParameters<Fn>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => void>(
      ctxAndFn: { context: Ctx; fn: Fn },
      ...args: effects.CpsFunctionParameters<Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(fork(...))`
   */
  fork: {
    <Fn extends (...args: any[]) => any>(fn: Fn, ...args: Parameters<Fn>): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctxAndFnName: [Ctx, Name],
      ...args: Parameters<Ctx[Name]>
    ): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctxAndFnName: { context: Ctx; fn: Name },
      ...args: Parameters<Ctx[Name]>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctxAndFn: [Ctx, Fn],
      ...args: Parameters<Fn>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctxAndFn: { context: Ctx; fn: Fn },
      ...args: Parameters<Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(spawn(...))`
   */
  spawn: {
    <Fn extends (...args: any[]) => any>(fn: Fn, ...args: Parameters<Fn>): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctxAndFnName: [Ctx, Name],
      ...args: Parameters<Ctx[Name]>
    ): R;

    <
      Ctx extends { [P in Name]: (this: Ctx, ...args: any[]) => any },
      Name extends string
    >(
      ctxAndFnName: { context: Ctx; fn: Name },
      ...args: Parameters<Ctx[Name]>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctxAndFn: [Ctx, Fn],
      ...args: Parameters<Fn>
    ): R;

    <Ctx, Fn extends (this: Ctx, ...args: any[]) => any>(
      ctxAndFn: { context: Ctx; fn: Fn },
      ...args: Parameters<Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(join(...))`
   */
  join: {
    (task: types.Task): R;

    (tasks: types.Task[]): R;
  };

  /**
   * Alias for `runner.should.yield(cancel(...))`
   */
  cancel: {
    (task: types.Task): R;

    (tasks: types.Task[]): R;

    (): R;
  };

  /**
   * Alias for `runner.should.yield(select(...))`
   */
  select: {
    (): R;

    <Fn extends (state: any, ...args: any[]) => any>(
      selector: Fn,
      ...args: effects.Tail<Parameters<Fn>>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(actionChannel(...))`
   */
  actionChannel: (
    pattern: types.ActionPattern,
    buffer?: types.Buffer<Action>,
  ) => R;

  /**
   * Alias for `runner.should.yield(flush(...))`
   */
  flush: <T>(channel: FlushableChannel<T>) => R;

  /**
   * Alias for `runner.should.yield(cancelled(...))`
   */
  cancelled: () => R;

  /**
   * Alias for `runner.should.yield(setContext(...))`
   */
  setContext: <C extends object>(props: C) => R;

  /**
   * Alias for `runner.should.yield(getContext(...))`
   */
  getContext: (prop: string) => R;

  /**
   * Alias for `runner.should.yield(delay(...))`
   */
  delay: <T = true>(ms: number, val?: T) => R;

  /**
   * Alias for `runner.should.yield(throttle(...))`
   */
  throttle: {
    <P extends types.ActionPattern>(
      ms: number,
      pattern: P,
      worker: (action: types.ActionMatchingPattern<P>) => any,
    ): R;

    <P extends types.ActionPattern, Fn extends (...args: any[]) => any>(
      ms: number,
      pattern: P,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<
        types.ActionMatchingPattern<P>,
        Fn
      >
    ): R;

    <A extends Action>(
      ms: number,
      pattern: types.ActionPattern<A>,
      worker: (action: A) => any,
    ): R;

    <A extends Action, Fn extends (...args: any[]) => any>(
      ms: number,
      pattern: types.ActionPattern<A>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<A, Fn>
    ): R;

    <T>(ms: number, channel: TakeableChannel<T>, worker: (item: T) => any): R;

    <T, Fn extends (...args: any[]) => any>(
      ms: number,
      channel: TakeableChannel<T>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<T, Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(debounce(...))`
   */
  debounce: {
    <P extends types.ActionPattern>(
      ms: number,
      pattern: P,
      worker: (action: types.ActionMatchingPattern<P>) => any,
    ): R;

    <P extends types.ActionPattern, Fn extends (...args: any[]) => any>(
      ms: number,
      pattern: P,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<
        types.ActionMatchingPattern<P>,
        Fn
      >
    ): R;

    <A extends Action>(
      ms: number,
      pattern: types.ActionPattern<A>,
      worker: (action: A) => any,
    ): R;

    <A extends Action, Fn extends (...args: any[]) => any>(
      ms: number,
      pattern: types.ActionPattern<A>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<A, Fn>
    ): R;

    <T>(ms: number, channel: TakeableChannel<T>, worker: (item: T) => any): R;

    <T, Fn extends (...args: any[]) => any>(
      ms: number,
      channel: TakeableChannel<T>,
      worker: Fn,
      ...args: effects.HelperWorkerParameters<T, Fn>
    ): R;
  };

  /**
   * Alias for `runner.should.yield(retry(...))`
   */
  retry: <Fn extends (...args: any[]) => any>(
    maxTries: number,
    delayLength: number,
    fn: Fn,
    ...args: Parameters<Fn>
  ) => R;

  /**
   * Alias for `runner.should.yield(all(...))`
   */
  all: {
    <T>(effects: T[]): R;

    <T>(effects: { [key: string]: T }): R;
  };

  /**
   * Alias for `runner.should.yield(race(...))`
   */
  race: {
    <T>(effects: { [key: string]: T }): R;

    <T>(effects: T[]): R;
  };
}
