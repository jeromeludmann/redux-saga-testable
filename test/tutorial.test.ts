import { Action } from 'redux';
import { select, put, call, take, delay } from 'redux-saga/effects';
import { createRunner, throwError, finalize } from '../src';

interface FetchUserAction extends Action<'FETCH_USER'> {
  payload: { userId: number };
}

interface FetchProductAction extends Action<'FETCH_PRODUCT'> {
  payload: { productId: number };
}

interface SendPingAction extends Action<'SEND_PING'> {}

const selectors = {
  getCurrentUser: (state: { user: any }) => state.user,
};

const services = {
  getUserById: (id: number) => {},
  getProductById: (id: number) => {},
  ping: () => {},
  notify: () => {},
};

function* fetchUserWorker(action: FetchUserAction) {
  const { userId } = action.payload;

  yield put({ type: 'FETCH_USER_REQUEST' });

  let user = yield select(selectors.getCurrentUser);
  if (user !== undefined) return;

  user = yield call(services.getUserById, userId);
  yield put({ type: 'FETCH_USER_SUCCESS', payload: user });
}

test('fetchUserWorker() should dispatch FETCH_USER_SUCCESS', () => {
  const userId = 123;
  const user = { user: 'name' };

  createRunner(fetchUserWorker, { type: 'FETCH_USER', payload: { userId } })
    .map(call(services.getUserById, userId), user)
    .should.put({ type: 'FETCH_USER_SUCCESS', payload: user });
});

test('fetchUserWorker() should not make the request if the user already exists', () => {
  const userId = 123;
  const existingUser = { user: 'name' };

  createRunner(fetchUserWorker, { type: 'FETCH_USER', payload: { userId } })
    .map(select(selectors.getCurrentUser), existingUser)
    .should.not.call(services.getUserById, userId);
});

test('fetchUserWorker() with your own assertions', () => {
  const userId = 123;
  const user = { user: 'name' };

  const output = createRunner(fetchUserWorker, {
    type: 'FETCH_USER',
    payload: { userId },
  })
    .map(call(services.getUserById, userId), user)
    .run();

  expect(output.effects).toHaveLength(4);

  expect(output.effects).toContainEqual(
    put({ type: 'FETCH_USER_SUCCESS', payload: user }),
  );
});

test('fetchUserWorker() with snapshot testing', () => {
  const userId = 123;
  const user = { user: 'name' };

  const output = createRunner(fetchUserWorker, {
    type: 'FETCH_USER',
    payload: { userId },
  })
    .map(call(services.getUserById, userId), user)
    .run();

  expect(output).toMatchSnapshot();
});

function* fetchProductWorker(action: FetchProductAction) {
  const { productId } = action.payload;

  try {
    yield put({ type: 'FETCH_PRODUCT_REQUEST' });
    const product = yield call(services.getProductById, productId);
    yield put({ type: 'FETCH_PRODUCT_SUCCESS', payload: product });
  } catch (error) {
    yield put({ type: 'FETCH_PRODUCT_FAILURE', payload: error.message });
  }
}

test('fetchProductWorker() should dispatch FETCH_PRODUCT_FAILURE if services.getProductById() fails', () => {
  const productId = 123;
  const error = new Error('Unable to fetch product');

  createRunner(fetchProductWorker, {
    type: 'FETCH_PRODUCT',
    payload: { productId },
  })
    .map(call(services.getProductById, productId), throwError(error))
    .should.put({ type: 'FETCH_PRODUCT_FAILURE', payload: error.message });
});

function* sendPingWorker(action: SendPingAction) {
  yield delay(1000);
  const pong1 = yield call(services.ping);

  yield delay(1000);
  const pong2 = yield call(services.ping);

  yield delay(1000);
  const pong3 = yield call(services.ping);

  yield put({
    type: 'RECEIVE_PONG',
    payload: { results: [pong1, pong2, pong3] },
  });
}

test('sendPingWorker() should dispatch RECEIVE_PONG with different results', () => {
  createRunner(sendPingWorker, { type: 'SEND_PING' })
    .map(call(services.ping), 12, 10, 11)
    .should.put({
      type: 'RECEIVE_PONG',
      payload: { results: [12, 10, 11] },
    });
});

function* notifyWatcher() {
  try {
    for (;;) {
      yield take('NOTIFY');
      yield put({ type: 'NOTIFY_REQUEST' });
      yield call(services.notify);
      yield put({ type: 'NOTIFY_SUCCESS' });
    }
  } finally {
    yield put({ type: 'NOTIFY_END' });
  }
}

test('notifyWatcher() should dispatch NOTIFY_END', () => {
  createRunner(notifyWatcher)
    .map(call(services.notify), finalize())
    .should.put({ type: 'NOTIFY_END' });
});

function* findProduct(id: number) {
  if (id < 0) {
    throw new Error(`Unable to find product ${id}`);
  }

  return yield call(services.getProductById, id);
}

test('findProduct() should throw an error when a negative id is given', () => {
  const id = -123;

  createRunner(findProduct, id)
    .catch(Error)
    .should.throw(/^Unable to find product/);
});
