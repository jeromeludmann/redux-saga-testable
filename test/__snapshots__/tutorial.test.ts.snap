// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchUserWorker() with snapshot testing 1`] = `
Object {
  "effects": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "type": "FETCH_USER_REQUEST",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": [MockFunction],
      },
      "type": "SELECT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          123,
        ],
        "context": null,
        "fn": [MockFunction],
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "payload": Object {
            "user": "name",
          },
          "type": "FETCH_USER_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "return": undefined,
}
`;
